src/http_clj/router.clj

```
@@ -1,26 +1,68 @@
 (ns http-clj.router
-  (:require [http-clj.request-handler :as handler]))
+  (:require [http-clj.request-handler :as handler])
+  (:import java.util.regex.Pattern))
 
 (defmulti path-matches?
-  (fn [_ route-path]
-    (type route-path)))
+  (fn [path route-path]
+    [(type path) (type route-path)]))
 
-(defmethod path-matches? String
-  [request-path route-path]
-  (= request-path route-path))
+(defmethod path-matches? [String String]
+  [path route-path]
+  (= path route-path))
 
-(defmethod path-matches? java.util.regex.Pattern
-  [request-path route-path]
-  (not (nil? (re-matches route-path request-path))))
+(defmethod path-matches? [String Pattern]
+  [path route-path]
+  (not (nil? (re-matches route-path path))))
 
-(defn- find-route [{path :path} routes]
+(defmethod path-matches? [Pattern Pattern]
+  [path route-path]
+  (= (.pattern path) (.pattern route-path)))
+
+(defmethod path-matches? [Pattern String]
```

I'm kind of surprised that this one is always `false` instead of doing the same thing as the `[String Pattern]` one does. If this case only happens in case of a programming error that should never happen, I'd sort of expect it to throw instead of just returning false.

---

I made it false because of situations like this

```clojure
(-> []
  (GET "/log" log-handler)
  (GET #"^/.*$" catch-all-handler))
```

The `GET` function uses the path to find a matching route in the routes vector. Making it false prevents  `(GET #"^/.*$" catch-all-handler)` from matching the `/log` route, since `#"^/.*$"` would match `/log` first. I wanted to limit regexes to only matching against regexes.

So it is expected that a regex will attempt to match against a string, which is why it would not be a programming error. If the regex and string were allowed to match in that direction, the resulting routes vector for the example above would be

```clojure
[{:path "/log" :handlers {"GET" catch-all-handler "HEAD" (partial head catch-all-handler}}]
```

vs.

```clojure
[{:path "/log" :handlers {"GET" log-handler "HEAD" (partial head log-handler)}}
 {:path #"^/.*$" :handlers {"GET" catch-all-handler "HEAD" (partial head catch-all-handler)}}]
```

This is complicated though. Maybe it would be simpler to limit paths to regexes instead of allowing strings or regexes? Thoughts?

---

And at the very least this deserves a comment, because I had to dig around to figure out/remember why I did it in the first place!

---

I guess my difficulty is that I don't know what the order of these two things *means*, even after looking at the code for the callers. For example, `path` sounds like it'd be the actual path in the URL, like `/damonkelley/http-clj/pull/10` for this PR's URL - but that must not be the case, since it's apparently possible for `path` to be a `java.util.regex.Pattern`. 

I took another 5-10 minutes to dig into the whole ns (https://github.com/damonkelley/http-clj/blob/0561bcbc5a7284bc2567ab68e8a497074e52fbbd/src/http_clj/router.clj) and now I see that at some point what was originally the `route-path` gets passed in as `path`. 

At some point along this namespace's progression, I think that `path-matches?` stuff ends up being reused for multiple reasons, and I think that's what is confusing me.

So I'm not sure, maybe there's some naming improvement that can help me understand here? Or maybe there's another way you can structure different routes' interactions to make `path-matches?` more single-purpose. For example, even now I'm still left with a question in my head: for all of these combinations, does the last matching one win? Or is it the first one?